x-pack/filebeat/input/httpjson/input.go

```diff
@@ -81,532 +49,150 @@ func newRetryLogger() *retryLogger {
-	case <-ticker.C:
-		in.log.Debug("Rate Limit: time is up.")
-		return nil
-	}
-}
-
-// createRequestInfoFromBody creates a new RequestInfo for a new HTTP request in pagination based on HTTP response body
-func createRequestInfoFromBody(config *Pagination, response, last common.MapStr, ri *RequestInfo) (*RequestInfo, error) {
-	// we try to get it from last element, if not found, from the original response
-	v, err := last.GetValue(config.IDField)
-	if err == common.ErrKeyNotFound {
-		v, err = response.GetValue(config.IDField)
-	}
-
-	if err == common.ErrKeyNotFound {
-		return nil, nil
+func newHTTPJSONInput(config config) (*httpJSONInput, error) {
+	if err := config.Validate(); err != nil {
+		return nil, err
 	}
 
+	tlsConfig, err := tlscommon.LoadTLSConfig(config.TLS)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to retrieve id_field for pagination")
+		return nil, err
 	}
 
-	if config.RequestField != "" {
-		ri.ContentMap.Put(config.RequestField, v)
-		if config.URL != "" {
-			ri.URL = config.URL
-		}
-	} else if config.URLField != "" {
-		url, err := url.Parse(ri.URL)
-		if err == nil {
-			q := url.Query()
-			q.Set(config.URLField, fmt.Sprint(v))
-			url.RawQuery = q.Encode()
-			ri.URL = url.String()
-		}
-	} else {
-		switch vt := v.(type) {
-		case string:
-			ri.URL = vt
-		default:
-			return nil, errors.New("pagination ID is not of string type")
-		}
-	}
-	if len(config.ExtraBodyContent) > 0 {
-		ri.ContentMap.Update(common.MapStr(config.ExtraBodyContent))
-	}
-	return ri, nil
+	return &httpJSONInput{
+		config:    config,
+		tlsConfig: tlsConfig,
+	}, nil
 }
 
-// processHTTPRequest processes HTTP request, and handles pagination if enabled
-func (in *HttpjsonInput) processHTTPRequest(ctx context.Context, client *http.Client, ri *RequestInfo) error {
-	ri.URL = in.getURL()
-
-	var (
-		m, v         interface{}
-		response, mm map[string]interface{}
-	)
+func (*httpJSONInput) Name() string { return inputName }
 
-	for {
-		req, err := in.createHTTPRequest(ctx, ri)
-		if err != nil {
-			return errors.Wrapf(err, "failed to create http request")
-		}
-		msg, err := client.Do(req)
-		if err != nil {
-			return errors.Wrapf(err, "failed to execute http client.Do")
-		}
-		responseData, err := ioutil.ReadAll(msg.Body)
-		header := msg.Header
-		msg.Body.Close()
-		if err != nil {
-			return errors.Wrapf(err, "failed to read http.response.body")
-		}
-		if msg.StatusCode != http.StatusOK {
-			in.log.Debugw("HTTP request failed", "http.response.status_code", msg.StatusCode, "http.response.body", string(responseData))
-			if msg.StatusCode == http.StatusTooManyRequests {
-				if err = in.applyRateLimit(ctx, header, in.config.RateLimit); err != nil {
-					return err
-				}
-				continue
-			}
-			return errors.Errorf("http request was unsuccessful with a status code %d", msg.StatusCode)
-		}
+func (in *httpJSONInput) Test(v2.TestContext) error {
+	url, err := url.Parse(in.config.URL)
+	if err != nil {
+		return err
+	}
```

Comment:
This is clearly a config parsing step and should therefore be done in the configure or newXInput function. Catch faulty configurations early. For URLs it would be nice to have a custom type that implements `Unpack(in string)`. All in all it would be nice to have a 'settings' package with reusable types that support Unpacking and Validate.